README.md
    tls::tls_init();

    let cfg = tls::TLSConfig::new().unwrap();
    cfg.set_key_file("./server.key").unwrap();
    cfg.set_cert_file("./server.crt").unwrap();

    let server = tls::TLS::server().unwrap();
    server.configure(&cfg).unwrap();

    let mut listener = std::io::TcpListener::bind("127.0.0.1:4433").unwrap()
        .listen().unwrap();

    std::thread::Thread::spawn(move || {
        let cfg = tls::TLSConfig::new().unwrap();
        cfg.insecure_noverifycert();
        cfg.insecure_noverifyhost();

        let mut client = tls::TLS::client().unwrap();
        client.configure(&cfg).unwrap();

        client.connect("127.0.0.1", "4433").unwrap();

        client.write(b"hello from client").unwrap();

        let buf:Vec<u8> = client.read_exact(17).unwrap();
        assert_eq!(buf, b"hello from server".to_vec());
    }).detach();

    let stream = listener.accept().unwrap();
    let mut conn = server.accept_socket(stream.as_raw_fd()).unwrap();

    conn.write(b"hello from server").unwrap();

    let buf = conn.read_exact(17).unwrap();
    assert_eq!(buf, b"hello from client".to_vec());

    conn.close().unwrap();


CHANGELOG
v0.0.2 (2015-01-04)
- Implement Reader and Writer


src/tls.rs
#![allow(dead_code)]

extern crate libc;

use bindings;
use std;

use std::c_str::ToCStr;
use std::sync::{Once, ONCE_INIT};

use std::io::{IoResult,IoError,IoErrorKind};

static INIT: Once = ONCE_INIT;

pub fn tls_init() {
	INIT.call_once(|| {
		unsafe { assert!(bindings::tls_init() == 0) }
	});
}

pub struct TLSConfig {
	ptr: *mut bindings::tls_config
}

impl Drop for TLSConfig {
	fn drop(&mut self) {
		unsafe { bindings::tls_config_free(self.ptr) }
	}
}

impl TLSConfig {
	pub fn new() -> Result<Self, ()> {
		let ptr = unsafe { bindings::tls_config_new() };

		if ptr.is_null() {
			Err(())
		}
		else {
			Ok(TLSConfig { ptr: ptr })
		}
	}

	fn set_string(&self, func: unsafe extern fn(*mut bindings::tls_config, *const i8) -> i32,
			string: &str) -> Result<(),()>
	{
		let res = string.with_c_str(|cstr|
			unsafe { func(self.ptr, cstr) }
		);
		if res == 0 {
			Ok(())
		}
		else {
			Err(())
		}
	}

	pub fn set_ca_file(&self, ca_file: &str) -> Result<(),()> {
		self.set_string(bindings::tls_config_set_ca_file, ca_file)
	}

	pub fn set_ca_path(&self, ca_path: &str) -> Result<(),()> {
		self.set_string(bindings::tls_config_set_ca_path, ca_path)
	}

	pub fn set_cert_file(&self, ca_path: &str) -> Result<(),()> {
		self.set_string(bindings::tls_config_set_cert_file, ca_path)
	}

	pub fn set_cert_mem(&self, cert: *const libc::c_int, len: libc::c_int)
			-> Result<(),()>
	{
		if cert.is_null() {
			return Err(())
		}

		let res = unsafe { bindings::tls_config_set_cert_mem(self.ptr, cert, len) };
		if res == 0 {
			Ok(())
		}
		else {
			Err(())
		}
	}

	pub fn set_ciphers(&self, ciphers: &str) -> Result<(),()> {
		self.set_string(bindings::tls_config_set_ciphers, ciphers)
	}

	pub fn set_ecdhcurve(&self, curve_name: &str) -> Result<(),()> {
		self.set_string(bindings::tls_config_set_ecdhcurve, curve_name)
	}

	pub fn set_key_file(&self, key_file: &str) -> Result<(),()> {
		self.set_string(bindings::tls_config_set_key_file, key_file)
	}

	pub fn set_key_mem(&self, key: *const libc::c_int, len: libc::c_int)
			-> Result<(),()>
	{
		if key.is_null() {
			return Err(())
		}

		let res = unsafe { bindings::tls_config_set_key_mem(self.ptr, key, len) };
		if res == 0 {
			Ok(())
		}
		else {
			Err(())
		}
	}

	pub fn set_protocols(&self, protocols: libc::c_int) {
		unsafe { bindings::tls_config_set_protocols(self.ptr, protocols) }
	}


	pub fn set_verify_depth(&self, verify_depth: libc::c_int) {
		unsafe { bindings::tls_config_set_verify_depth(self.ptr, verify_depth) }
	}

	pub fn clear_keys(&self) {
		unsafe { bindings::tls_config_clear_keys(self.ptr) }
	}

	pub fn insecure_noverifyhost(&self) {
		unsafe { bindings::tls_config_insecure_noverifyhost(self.ptr) }
	}

	pub fn insecure_noverifycert(&self) {
		unsafe { bindings::tls_config_insecure_noverifycert(self.ptr) }
	}

	pub fn verify(&self) {
		unsafe { bindings::tls_config_verify(self.ptr) }
	}
}


pub struct TLS {
	ptr: *mut bindings::tls
}

impl Drop for TLS {
	fn drop(&mut self) {
		unsafe { bindings::tls_free(self.ptr) }
	}
}

impl TLS {
	pub fn client() -> Result<TLS,()> {
		let ptr = unsafe { bindings::tls_client() };
		if ptr.is_null() {
			Err(())
		} else {
			Ok(TLS { ptr: ptr })
		}
	}

	pub fn server() -> Result<TLS,()> {
		let ptr = unsafe { bindings::tls_server() };
		if ptr.is_null() {
			Err(())
		} else {
			Ok(TLS { ptr: ptr })
		}
	}

	pub fn configure(&self, config: &TLSConfig) -> Result<(),&'static str> {
		let res = unsafe { bindings::tls_configure(self.ptr, config.ptr) };

		if res == 0 {
			Ok(())
		}
		else {
			Err(self.get_error())
		}
	}

	fn get_error(&self) -> &'static str {
		unsafe {
			let s = bindings::tls_error(self.ptr);
			std::str::from_c_str(s)
		}
	}

	pub fn reset(&self) {
		unsafe { bindings::tls_reset(self.ptr) }
	}

	pub fn accept_socket(&self, socket: libc::c_int) -> Result<TLS,&'static str> {
		let mut ptr = 0 as *mut bindings::tls;
		let res = unsafe { bindings::tls_accept_socket(self.ptr, &mut ptr, socket) };

		if res != 0 || ptr.is_null() {
			Err(self.get_error())
		}
		else {
			Ok(TLS { ptr:ptr })
		}
	}

	pub fn connect(&self, host: &str, port: &str) -> Result<(), &'static str> {
		host.with_c_str(|host_ptr| {
			port.with_c_str(|port_ptr| {
				let res = unsafe { bindings::tls_connect(self.ptr, host_ptr, port_ptr) };

				if res == 0 {
					Ok(())
				} else {
					Err(self.get_error())
				}
			})
		})
	}

	pub fn connect_fds(&self, fd_read: libc::c_int, fd_write: libc::c_int, hostname: &str)
			-> Result<(), &'static str>
	{
		hostname.with_c_str(|host| {
			let res = unsafe { bindings::tls_connect_fds(self.ptr, fd_read, fd_write, host) };

			if res == 0 {
				Ok(())
			} else {
				Err(self.get_error())
			}
		})
	}

	pub fn connect_socket(&self, sock: libc::c_int, hostname: &str) -> Result<(), &'static str> {
		hostname.with_c_str(|host| {
			let res = unsafe { bindings::tls_connect_socket(self.ptr, sock, host) };

			if res == 0 {
				Ok(())
			} else {
				Err(self.get_error())
			}
		})
	}
	pub fn close(&self) -> Result<(), &'static str> {
		let res = unsafe { bindings::tls_close(self.ptr) };

		if res == 0 {
			Ok(())
		} else {
			Err(self.get_error())
		}
	}
}

impl Reader for TLS {
	fn read(&mut self, buf: &mut [u8]) -> IoResult<uint> {
		let mut outlen = 0 as libc::size_t;

		let res = unsafe {
			bindings::tls_read(self.ptr, buf.as_mut_ptr() as *mut libc::c_void,
				buf.len() as libc::c_int, &mut outlen)
		};
		
		match res {
			0 => {
				Ok(outlen as uint)
			},
			bindings::TLS_READ_AGAIN => Ok(0),
			bindings::TLS_WRITE_AGAIN => Ok(0),
			6 /*SSL_ERROR_ZERO_RETURN*/ => Err(IoError {
				kind: IoErrorKind::EndOfFile,
				desc: self.get_error(),
				detail: None,
			}),
			_ => Err(IoError {
				kind: IoErrorKind::OtherIoError,
				desc: self.get_error(),
				detail: None,
			}),
		}
	}
}

impl Writer for TLS {
	fn write(&mut self, buf: &[u8]) -> IoResult<()> {
		let mut outlen = 0 as libc::size_t;

		let res = unsafe {
			bindings::tls_write(self.ptr, buf.as_ptr() as *const libc::c_void,
				buf.len() as i32, &mut outlen)
		};
		
		match res {
			0 => Ok(()),
			bindings::TLS_READ_AGAIN => Err(IoError {
				kind: IoErrorKind::EndOfFile,
				desc: "READ_AGAIN",
				detail: None,
			}),
			bindings::TLS_WRITE_AGAIN => Err(IoError {
				kind: IoErrorKind::EndOfFile,
				desc: "WRITE_AGAIN",
				detail: None,
			}),
			_ => Err(IoError {
				kind: IoErrorKind::OtherIoError,
				desc: self.get_error(),
				detail: None,
			}),
		}
	}
}

src/bindings.rs
extern crate libc;

/*
int tls_init()
*/
#[link(name="tls")]
#[link(name="crypto")]
#[link(name="ssl")]
extern "C" {
	pub fn tls_init() -> libc::c_int;
}


/*
const char * tls_error()
	(struct tls *) ctx [struct tls *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_error(ctx: *mut tls) -> *const libc::c_char;
}


/*
struct tls_config * tls_config_new() [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_new() -> *mut tls_config;
}


/*
void tls_config_free()
	(struct tls_config *) config [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_free(config: *mut tls_config);
}


/*
int tls_config_set_ca_file()
	(struct tls_config *) config [struct tls_config *]
	(const char *) ca_file
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_ca_file(config: *mut tls_config, ca_file: *const libc::c_char) -> libc::c_int;
}


/*
int tls_config_set_ca_path()
	(struct tls_config *) config [struct tls_config *]
	(const char *) ca_path
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_ca_path(config: *mut tls_config, ca_path: *const libc::c_char) -> libc::c_int;
}


/*
int tls_config_set_cert_file()
	(struct tls_config *) config [struct tls_config *]
	(const char *) cert_file
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_cert_file(config: *mut tls_config, cert_file: *const libc::c_char) -> libc::c_int;
}


/*
int tls_config_set_cert_mem()
	(struct tls_config *) config [struct tls_config *]
	(const int *) cert
	(int) len
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_cert_mem(config: *mut tls_config, cert: *const libc::c_int, len: libc::c_int) -> libc::c_int;
}


/*
int tls_config_set_ciphers()
	(struct tls_config *) config [struct tls_config *]
	(const char *) ciphers
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_ciphers(config: *mut tls_config, ciphers: *const libc::c_char) -> libc::c_int;
}


/*
int tls_config_set_ecdhcurve()
	(struct tls_config *) config [struct tls_config *]
	(const char *) name
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_ecdhcurve(config: *mut tls_config, name: *const libc::c_char) -> libc::c_int;
}


/*
int tls_config_set_key_file()
	(struct tls_config *) config [struct tls_config *]
	(const char *) key_file
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_key_file(config: *mut tls_config, key_file: *const libc::c_char) -> libc::c_int;
}


/*
int tls_config_set_key_mem()
	(struct tls_config *) config [struct tls_config *]
	(const int *) key
	(int) len
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_key_mem(config: *mut tls_config, key: *const libc::c_int, len: libc::c_int) -> libc::c_int;
}


/*
void tls_config_set_protocols()
	(struct tls_config *) config [struct tls_config *]
	(int) protocols
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_protocols(config: *mut tls_config, protocols: libc::c_int);
}


/*
void tls_config_set_verify_depth()
	(struct tls_config *) config [struct tls_config *]
	(int) verify_depth
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_set_verify_depth(config: *mut tls_config, verify_depth: libc::c_int);
}


/*
void tls_config_clear_keys()
	(struct tls_config *) config [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_clear_keys(config: *mut tls_config);
}


/*
void tls_config_insecure_noverifyhost()
	(struct tls_config *) config [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_insecure_noverifyhost(config: *mut tls_config);
}


/*
void tls_config_insecure_noverifycert()
	(struct tls_config *) config [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_insecure_noverifycert(config: *mut tls_config);
}


/*
void tls_config_verify()
	(struct tls_config *) config [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_config_verify(config: *mut tls_config);
}


/*
struct tls * tls_client() [struct tls *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_client() -> *mut tls;
}


/*
struct tls * tls_server() [struct tls *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_server() -> *mut tls;
}


/*
int tls_configure()
	(struct tls *) ctx [struct tls *]
	(struct tls_config *) config [struct tls_config *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_configure(ctx: *mut tls, config: *mut tls_config) -> libc::c_int;
}


/*
void tls_reset()
	(struct tls *) ctx [struct tls *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_reset(ctx: *mut tls);
}


/*
void tls_free()
	(struct tls *) ctx [struct tls *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_free(ctx: *mut tls);
}


/*
int tls_accept_socket()
	(struct tls *) ctx [struct tls *]
	(struct tls **) cctx [struct tls **]
	(int) socket
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_accept_socket(ctx: *mut tls, cctx: *mut *mut tls, socket: libc::c_int) -> libc::c_int;
}


/*
int tls_connect()
	(struct tls *) ctx [struct tls *]
	(const char *) host
	(const char *) port
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_connect(ctx: *mut tls, host: *const libc::c_char, port: *const libc::c_char) -> libc::c_int;
}


/*
int tls_connect_fds()
	(struct tls *) ctx [struct tls *]
	(int) fd_read
	(int) fd_write
	(const char *) hostname
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_connect_fds(ctx: *mut tls, fd_read: libc::c_int, fd_write: libc::c_int, hostname: *const libc::c_char) -> libc::c_int;
}


/*
int tls_connect_socket()
	(struct tls *) ctx [struct tls *]
	(int) s
	(const char *) hostname
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_connect_socket(ctx: *mut tls, s: libc::c_int, hostname: *const libc::c_char) -> libc::c_int;
}


/*
int tls_read()
	(struct tls *) ctx [struct tls *]
	(void *) buf
	(int) buflen
	(size_t *) outlen
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_read(ctx: *mut tls, buf: *mut libc::c_void, buflen: libc::c_int, outlen: *mut libc::size_t) -> libc::c_int;
}


/*
int tls_write()
	(struct tls *) ctx [struct tls *]
	(const void *) buf
	(int) buflen
	(size_t *) outlen
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_write(ctx: *mut tls, buf: *const libc::c_void, buflen: libc::c_int, outlen: *mut libc::size_t) -> libc::c_int;
}


/*
int tls_close()
	(struct tls *) ctx [struct tls *]
*/
#[link(name="tls")]
extern "C" {
	pub fn tls_close(ctx: *mut tls) -> libc::c_int;
}


/*
struct tls
*/
#[repr(C)]
pub struct tls;

/*
struct tls_config
*/
#[repr(C)]
pub struct tls_config;

/* HEADER_TLS_H # */

/* TLS_API 20141031 # */
pub const TLS_API: i32 = 20141031;

/* TLS_PROTOCOL_TLSv1_0 ( 1 << 1 ) # */

/* TLS_PROTOCOL_TLSv1_1 ( 1 << 2 ) # */

/* TLS_PROTOCOL_TLSv1_2 ( 1 << 3 ) # */

/* TLS_PROTOCOL_TLSv1 ( TLS_PROTOCOL_TLSv1_0 | TLS_PROTOCOL_TLSv1_1 | TLS_PROTOCOL_TLSv1_2 ) # */

/* TLS_PROTOCOLS_DEFAULT TLS_PROTOCOL_TLSv1 # */

/* TLS_READ_AGAIN - 2 # */
pub const TLS_READ_AGAIN: i32 = -2;

/* TLS_WRITE_AGAIN - 3 struct */
pub const TLS_WRITE_AGAIN: i32 = -3;



src/lib.rs
#![allow(unused_imports)] 
#![allow(dead_code)] 

use std::io::{TcpListener, TcpStream};
use std::io::{Acceptor, Listener};
use std::os::unix::prelude::AsRawFd;

mod bindings;
mod tls;

#[test]
fn it_works() {
	tls::tls_init();

	let cfg = tls::TLSConfig::new().unwrap();
	cfg.set_key_file("./server.key").unwrap();
	cfg.set_cert_file("./server.crt").unwrap();

	let server = tls::TLS::server().unwrap();
	server.configure(&cfg).unwrap();

	let mut listener = std::io::TcpListener::bind("127.0.0.1:4433").unwrap()
		.listen().unwrap();

	std::thread::Thread::spawn(move || {
		let cfg = tls::TLSConfig::new().unwrap();
		cfg.insecure_noverifycert();
		cfg.insecure_noverifyhost();

		let mut client = tls::TLS::client().unwrap();
		client.configure(&cfg).unwrap();

		client.connect("127.0.0.1", "4433").unwrap();

		client.write(b"hello from client").unwrap();

		let buf:Vec<u8> = client.read_exact(17).unwrap();
		assert_eq!(buf, b"hello from server".to_vec());
	}).detach();

	let stream = listener.accept().unwrap();
	let mut conn = server.accept_socket(stream.as_raw_fd()).unwrap();

	conn.write(b"hello from server").unwrap();

	let buf = conn.read_exact(17).unwrap();
	assert_eq!(buf, b"hello from client".to_vec());

	conn.close().unwrap();
}
